### PR TITLE
Setting up allure for integration tests

### PR DESCRIPTION
**Pull request by swetha1654**
### Overview

Integrated [Allure tool ](https://allurereport.org/docs/)for visualization of integration tests.

### Rationale

This integration is required so that we can analyse the reasons for failure of integration tests better.

### Workflow Changes

- The pytest collection plugin is defined under `python/pytest-plugins` for setting the default result status as unknown for all tests during collection time.
- The `allure-add-default-for-missing-results` is defined to merge the results obtained from the collection time with the actual results.
- Changes were made to integration-test-run.yaml to introduce the collection of tests and use the pytest plugin then.
- Allure report generation workflow was defined to combine the results, create a report and upload it to **gh-pages**

### Checklist

- [x] The [contributing guide](https://github.com/canonical/is-charms-contributing-guide) was applied
- [x] The PR is tagged with appropriate label (`urgent`, `trivial`, `complex`)

### Results

Here's an [example report](https://canonical.github.io/github-runner-operator/1593/#) from github-runners repo using this workflow.

### CODE REVIEW

File: .github/workflows/allure_report.yaml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+          pattern: allure-default-results*
+          merge-multiple: true
+      - name: Download actual test results
+        uses: actions/download-artifact@v4
+        with:
+          path: allure-results/
+          pattern: allure-results*
+          merge-multiple: true
+      - name: Install CLI
+        run: pipx install git+https://github.com/canonical/operator-workflows@ISD-2620-allure#subdirectory=python/cli
```
Comment (swetha1654) [inline .github/workflows/allure_report.yaml:42]: will change all branch names from ISD-2620-allure back to main once i get one round of reviews

File: .github/workflows/allure_report.yaml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        run: |
+          if [[ -d repo/_latest/history/ ]]
+          then
+            echo 'Loading history'
+            cp -r repo/_latest/history/ allure-results/
+          fi
+      - name: Create executor.json
+        shell: python
+        run: |
+          # Reverse engineered from https://github.com/simple-elf/allure-report-action/blob/eca283b643d577c69b8e4f048dd6cd8eb8457cfd/entrypoint.sh
```
Comment (jdkandersson) [inline .github/workflows/allure_report.yaml:58]: Could we use the action?

File: .github/workflows/allure_report.yaml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+          DATA = f"""<!DOCTYPE html>
+          <meta charset="utf-8">
+          <meta http-equiv="cache-control" content="no-cache">
+          <meta http-equiv="refresh" content="0; url=${{ github.run_number }}">
+          """
+          with open("repo/index.html", "w") as file:
+              file.write(DATA)
+      - name: Update GitHub pages branch
+        working-directory: repo/
+        # TODO future improvement: commit message
```
Comment (jdkandersson) [inline .github/workflows/allure_report.yaml:85]: Instead of TODO please make it a story on our backlog

File: .github/workflows/allure_report.yaml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+          mkdir '${{ github.run_number }}'
+          rm -f _latest
+          ln -s '${{ github.run_number }}' _latest
+          cp -r ../allure-report/. _latest/
+          git add .
+          git config user.name "GitHub Actions"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
+          git commit -m "Allure report ${{ github.run_number }}"
+          # Uses token set in checkout step
+          git push origin gh-pages
```
Comment (jdkandersson) [inline .github/workflows/allure_report.yaml:96]: Final new line please 🙂

File: .github/workflows/integration_test_allure.md
Comment (jdkandersson) [inline .github/workflows/integration_test_allure.md:1]: Is this the documentation? It should probably go somewhere else like in the `docs` directory or something? Perhaps ping @erinecon to see where it should go

File: python/cli/README.md
Comment (jdkandersson) [inline python/cli/README.md:1]: What is this please?

Comment (jdkandersson) [reply python/cli/README.md:1]: Could we just install it from their repo actually?

File: .github/workflows/allure_report.yaml [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+              "reportUrl": "../${{ github.run_number }}/",
+          }
+          with open("allure-results/executor.json", "w") as file:
+              json.dump(DATA, file)
+      - name: Generate Allure report
+        run: allure generate
+      - name: Create index.html
+        shell: python
+        run: |
+          DATA = f"""<!DOCTYPE html>
```
Comment (yanksyoon) [inline .github/workflows/allure_report.yaml:76]: Would this require textwrap.dedent?

File: .github/workflows/allure_report.yaml (hunk 4 above)
Comment (yanksyoon) [inline .github/workflows/allure_report.yaml:96]: I think a newline is missing here :)

File: .github/workflows/allure_report.yaml [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+          """
+          with open("repo/index.html", "w") as file:
+              file.write(DATA)
+      - name: Update GitHub pages branch
+        working-directory: repo/
+        # TODO future improvement: commit message
+        run: |
+          mkdir '${{ github.run_number }}'
+          rm -f _latest
+          ln -s '${{ github.run_number }}' _latest
```
Comment (yanksyoon) [inline .github/workflows/allure_report.yaml:89]: Is there a reason for a soft link here?

File: .github/workflows/allure_report.yaml [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+          cp -r ../allure-report/. _latest/
+          git add .
+          git config user.name "GitHub Actions"
+          git config user.email "41898282+github-actions[bot]@users.noreply.github.com"
```
Comment (yanksyoon) [inline .github/workflows/allure_report.yaml:93]: Would a comment explaining where this user came from (https://github.com/actions/checkout/issues/13#issuecomment-724415212) help?

File: .github/workflows/integration_test_allure.md [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+Here's an [example for the above](https://github.com/canonical/github-runner-operator/pull/412).
+
+**NOTE:** If the workflow is being called inside a matrix with the same test modules run with different parameters, the allure report will only display the results of the last combination.
+
+## 3. Changing branch permissions
+
+**NOTE:** For this, you would require admin access to the repository.
+
+- Go to the repository's **Settings > Branches** and next to Branch protection rules, select **Add rule**
+- Enter the branch name **gh-pages** and select **Allow force pushes** and click **Save changes**
```
Comment (yanksyoon) [inline .github/workflows/integration_test_allure.md:42]: I think our repository rules by default do not allow force pushes, would this be required? Just wondering if we can prevent force pushes:)

File: .github/workflows/integration_test_allure.md [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+
+## 4. Github pages branch
+
+- Create `gh-pages` branch:
+
+```
```
Comment (yanksyoon) [inline .github/workflows/integration_test_allure.md:48]: Could this segment be automated as well so the repos can start using this right away?

File: .github/workflows/integration_test_run.yaml [hunk 10]
```diff
@@ -284,26 +312,23 @@ jobs:
...
-          series=""
-          if [ ! -z ${{ matrix.series }} ]; then
-            series="--series ${{ matrix.series }}"
-          fi
-          echo "SERIES=$series" >> $GITHUB_ENV
-          module=""
-          if [ ! -z ${{ matrix.modules }} ]; then
-            module="-k ${{ matrix.modules }}"
-          fi
-          echo "MODULE=$module" >> $GITHUB_ENV
```
Comment (yanksyoon) [inline .github/workflows/integration_test_run.yaml:296]: Let me just double check that this doesnt impact the existing workflows :)

File: python/cli/README.md [hunk 11]
```diff
@@ -0,0 +1,3 @@
+The `allure_add_default_for_missing_results` cli combines the Allure default results & actual results before generating the report. 
+
+Currently, if the result of a test is not available (for reasons such as setup failure), allure omits the test. This CLI ensures that for every test, if actual result is available, it will use that. Otherwise, it uses the default result `unknown`.
```
Comment (yanksyoon) [inline python/cli/README.md:3]: Missing newline here as well:)

File: python/cli/allure_add_default_for_missing_results/allure_add_default_for_missing_results.py
Comment (yanksyoon) [inline python/cli/allure_add_default_for_missing_results/allure_add_default_for_missing_results.py:1]: Would these files be from data team? Any way we could reuse their workflow? (In order to have single source of truth)

**Review by yanksyoon — COMMENTED** [review]
LGTM!
Minor nits :)

As of newlines, https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline i think this might help reason about why its required:D

File: .github/workflows/allure_report.yaml (hunk 6 above)
Comment (swetha1654) [reply .github/workflows/allure_report.yaml:89]: the symbolic link is used in the next run to create the [history](https://github.com/canonical/data-platform-workflows/blob/8ec5cc7242d90d7ae1e125755297a97402e54584/.github/workflows/integration_test_charm.yaml#L579-L580) (in the allure reports, it shows the last 20 runs in some graphs) 
we store each report in a separate directory so that it's possible to view older reports in addition to the latest report

File: .github/workflows/allure_report.yaml (hunk 7 above)
Comment (swetha1654) [reply .github/workflows/allure_report.yaml:93]: addressed

File: .github/workflows/allure_report.yaml (hunk 4 above)
Comment (swetha1654) [reply .github/workflows/allure_report.yaml]: addressed

File: python/cli/README.md
Comment (swetha1654) [reply python/cli/README.md:1]: addressed

File: .github/workflows/allure_report.yaml (hunk 4 above)
Comment (swetha1654) [reply .github/workflows/allure_report.yaml]: addressed

File: python/cli/README.md (hunk 11 above)
Comment (swetha1654) [reply python/cli/README.md]: removed the README as i'm using the cli from data platforms repo directly

File: docs/how-to/integration_test_allure.md [hunk 12]
```diff
@@ -0,0 +1,64 @@
+# Allure Report Integration
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:1]: Since this is now a how-to guide, I would rephrase the title in the style of `# How to...`. Maybe something like `# How to set up Allure Reports for integration tests` -- let me know if you have any thoughts or ideas :)

File: docs/how-to/integration_test_allure.md [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+
+[Allure Report](https://allurereport.org/) for [integration_test_run.yaml](https://github.com/canonical/operator-workflows?tab=readme-ov-file#integration-test-workflow-canonicaloperator-workflowsgithubworkflowsintegration_testyamlmain)
+
+In order to integrate Allure with your repository, perform the following actions:
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:5]: I would combine these two lines into a single sentence that describes what the how-to guide will talk about. Maybe something like, `This how-to guide describes how to integrate [Allure Reports](https://allurereport.org/) into your code repository's [integration_test_run.yaml](https://github.com/canonical/operator-workflows?tab=readme-ov-file#integration-test-workflow-canonicaloperator-workflowsgithubworkflowsintegration_testyamlmain).`

File: docs/how-to/integration_test_allure.md [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+
+## 1. Adding allure-pytest and pytest collection plugin
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:7]:
```suggestion
## Add allure-pytest and pytest collection plugin
```

File: docs/how-to/integration_test_allure.md [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+
+Please add the following into the `requirements.txt` that is called by the integration test -
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:9]:
```suggestion
Include the following snippet in the `requirements.txt` that is called by the integration test:
```

File: docs/how-to/integration_test_allure.md [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+
+```
+allure-pytest>=2.8.18
+```
+
+Add the following line under the dependencies (`deps`) in the integration section inside `tox.ini` -
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:15]:
```suggestion
Add the following line under the dependencies (`deps`) in the integration section inside `tox.ini`:
```

File: docs/how-to/integration_test_allure.md [hunk 17]
```diff
@@ -0,0 +1,64 @@
...
+
+```
+git+https://github.com/canonical/operator-workflows@main\#subdirectory=python/pytest_plugins/allure_pytest_collection_report
+```
+
+## 2. Calling the allure-workflow
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:21]:
```suggestion
## Call the allure-workflow
```

File: docs/how-to/integration_test_allure.md [hunk 18]
```diff
@@ -0,0 +1,64 @@
...
+
+Add the following lines at the end of the workflow that runs the integrations tests by calling the reuable workflow [integration-test.yaml](https://github.com/canonical/operator-workflows/blob/main/.github/workflows/integration_test.yaml):
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:23]:
```suggestion
To call the reusable workflow [integration-test.yaml](https://github.com/canonical/operator-workflows/blob/main/.github/workflows/integration_test.yaml), add the following lines at the end of the workflow that runs the integrations tests:
```

File: docs/how-to/integration_test_allure.md [hunk 19]
```diff
@@ -0,0 +1,64 @@
...
+  allure-report:
+    if: always() && !cancelled()
+    needs:
+      - [list of jobs with tests you would like to visualize]
+    uses: canonical/operator-workflows/.github/workflows/allure_report.yaml@main
+```
+
+Here's an [example for the above](https://github.com/canonical/github-runner-operator/pull/412).
+
+**NOTE:** If the workflow is being called inside a matrix with the same test modules run with different parameters, the allure report will only display the results of the last combination.
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:35]:
```suggestion
**NOTE:** If the workflow is being called inside a matrix with the same test modules run with different parameters, the Allure Report will only display the results of the last combination.
```

File: docs/how-to/integration_test_allure.md [hunk 20]
```diff
@@ -0,0 +1,64 @@
...
+
+## 3. Changing branch permissions
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:37]:
```suggestion
## Change branch permissions
```

File: docs/how-to/integration_test_allure.md [hunk 21]
```diff
@@ -0,0 +1,64 @@
...
+
+**NOTE:** For this, you would require admin access to the repository.
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:39]:
```suggestion
**NOTE:** For this step, you need admin access to the repository.
```

File: docs/how-to/integration_test_allure.md [hunk 22]
```diff
@@ -0,0 +1,64 @@
...
+
+- Go to the repository's **Settings > Branches** and next to Branch protection rules, select **Add rule**
+- Enter the branch name **gh-pages** and click **Save changes** (Ensure that "require signed commits" is unchecked)
+
+## 4. Github pages branch
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:44]:
```suggestion
## Create a Github pages branch
```

File: docs/how-to/integration_test_allure.md [hunk 23]
```diff
@@ -0,0 +1,64 @@
...
+
+```
+  allure-report:
+    if: always() && !cancelled()
+    needs:
+      - [list of jobs with tests you would like to visualize]
+    uses: canonical/operator-workflows/.github/workflows/allure_report.yaml@main
+```
+
+Here's an [example for the above](https://github.com/canonical/github-runner-operator/pull/412).
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:33]: Nitpick: I would rephrase this a little differently. Something like `For an example of this implementation, see [the GitHub runner repository](https://github.com/canonical/github-runner-operator/pull/412).`

File: docs/how-to/integration_test_allure.md [hunk 24]
```diff
@@ -0,0 +1,64 @@
...
+ # git checkout --orphan gh-pages
+ # git rm -rf . 
+ # touch .nojekyll 
+ # git add .nojekyll 
+ # git commit -m "Initial commit" 
+ # git push origin gh-pages
+ # ) 
+ ```
+
+ - Enable gh pages publishing at ** Settings > Pages ** and set branch name as `gh-pages`:
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:60]: Does `gh` stand for GitHub? If so, I would explicitly write `GitHub` here

File: docs/how-to/integration_test_allure.md [hunk 25]
```diff
@@ -0,0 +1,64 @@
...
+
+<img width="816" alt="image" src="https://github.com/user-attachments/assets/346c04fc-0daa-40bc-92b5-93b0ea639f94">
+
+ [Example PR for steps 1 & 2](https://github.com/canonical/github-runner-operator/pull/412/files#)
```
Comment (erinecon) [inline docs/how-to/integration_test_allure.md:64]: Nitpick: I would rephrase this a little differently. Something like `For an example of the first two steps, see [the GitHub runner repository](https://github.com/canonical/github-runner-operator/pull/412).`

**Review by erinecon — COMMENTED** [review]
Thanks for writing this how-to guide! My review is mostly nitpicks to make the style checker happy :)

File: docs/how-to/integration_test_allure.md (hunk 12 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: addressed

File: docs/how-to/integration_test_allure.md (hunk 13 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: addressed

File: docs/how-to/integration_test_allure.md (hunk 14 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: addressed

File: docs/how-to/integration_test_allure.md (hunk 15 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: addressed

File: docs/how-to/integration_test_allure.md (hunk 16 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: addressed

File: docs/how-to/integration_test_allure.md (hunk 17 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: addressed

File: docs/how-to/integration_test_allure.md (hunk 18 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md:21]: changed this sentence slightly, the meaning changes when its rephrased like this, i have put allure-report link instead.

File: docs/how-to/integration_test_allure.md (hunk 19 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md:33]: addressed

File: docs/how-to/integration_test_allure.md (hunk 20 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: done

File: docs/how-to/integration_test_allure.md (hunk 21 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: addressed

File: docs/how-to/integration_test_allure.md (hunk 23 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: done

File: docs/how-to/integration_test_allure.md (hunk 24 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: done

File: docs/how-to/integration_test_allure.md (hunk 25 above)
Comment (swetha1654) [reply docs/how-to/integration_test_allure.md]: done

File: .github/workflows/allure_report.yaml (hunk 2 above)
Comment (jdkandersson) [reply .github/workflows/allure_report.yaml:60]: Ok sounds fine, perhaps include this as a reason as a comment

**Review by jdkandersson — COMMENTED** [review]
Looks fine, please get a review from @weiiwang01 as the senior review in this case

**Comment by jdkandersson** [comment]
Thank you for adding this, I think it will help us improve our tests a lot!

File: .github/workflows/allure_report.yaml (hunk 2 above)
Comment (swetha1654) [reply .github/workflows/allure_report.yaml:60]: sure

File: .github/workflows/integration_test_allure.md (hunk 9 above)
Comment (yanksyoon) [reply .github/workflows/integration_test_allure.md]: I believe the separate protection rule is no longer required? Hence we could check if the branch is present, continue if it is, create one if it isn't :)

Comment (swetha1654) [reply .github/workflows/integration_test_allure.md]: by default for all the repos we have a global protection rule for all branches requiring signed commits. We require a seperate protection rule for gh-pages branches with the signed commits unchecked otherwise the last step in the allure_report workflow fails to push the changes into that branch.

Comment (swetha1654) [reply .github/workflows/integration_test_allure.md]: image
this is the global rule i was referring to

File: .github/workflows/integration_test_run.yaml [hunk 26]
```diff
@@ -171,6 +171,36 @@ jobs:
...
+          if [ ! -z ${{ matrix.modules }} ]; then
+            module="-k ${{ matrix.modules }}"
+            allure_artifact_suffix="$allure_artifact_suffix"-${{ matrix.modules }}
+          fi
+          echo "MODULE=$module" >> $GITHUB_ENV
+          echo "ALLURE_ARTIFACT_SUFFIX=$allure_artifact_suffix" >> $GITHUB_ENV
+      - name: Install tox
+        run: |
+          pip install tox 
+      - name: Collect tests for Allure
```
Comment (weiiwang01) [inline .github/workflows/integration_test_run.yaml:194]: I think maybe we can add a step before this one to check if the Allure as a dependency is declared for this tox test environment and toggle the Allure-related steps on and off accordingly. This way, for projects without Allure installed, it won't run Allure or break the workflow.

For example, something like this (untested):
```
- name: Check Allure
  run: |
    tox -e ${{ inputs.test-tox-env }} --notest --list-dependencies 2>&1 | grep -q allure \
      && echo 'ENABLE_ALLURE=true' >> $GITHUB_ENV \
      || echo 'ENABLE_ALLURE=false' >> $GITHUB_ENV
```
Then use `if: ${{ env.ENABLE_ALLURE == "true" }}` in the following Allure-related steps.

Comment (swetha1654) [reply .github/workflows/integration_test_run.yaml:225]: addressed

**Review by yanksyoon — APPROVED** [review]

**Review by weiiwang01 — APPROVED** [review]
